### PR TITLE
[Serialization] Lazily Resolve the Eraser Type of @ _typeEraser

### PR DESCRIPTION
**Pull request by CodaFi**
Type erasure requires a circular construction by its very nature:

```
@_typeEraser(AnyProto)
protocol Proto { /**/ }
public struct AnyProto : Proto {}
```

If we eagerly resolve AnyProto, the chain of resolution steps that
deserialization must make goes a little something like this:

```
Lookup(Proto)
    -> Deserialize(@_typeEraser(AnyProto))
    -> Lookup(AnyProto)
    -> DeserializeInheritedStuff(AnyProto)
    -> Lookup(Proto)
```

Now, this cycle could be broken if the order of incremental inputs was
such that we had already cached the lookup of Proto.

Resolve this cycle in any case by suspending the deserialization of the
type eraser until the point it's demanded by adding
ResolveTypeEraserTypeRequest.

rdar://61270195

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci smoke test

File: include/swift/AST/Attr.h [hunk 1]
```diff
@@ -1111,16 +1111,58 @@ class DynamicReplacementAttr final
...
+  }
+
+  /// Returns \c true if the type eraser type has a valid implementation of the
+  /// erasing initializer for the given protocol.
   bool hasViableTypeEraserInit(ProtocolDecl *protocol) const;
 
+  /// Resolves the type of this attribute.
+  ///
+  /// This entrypoint is suitable for semantic clients like the
+  /// expression checker. Syntactic clients shuld use
```
Comment (CodaFi) [inline include/swift/AST/Attr.h:1162]: I coulda sworn I learned this language first...

**Review by hborla — APPROVED** [review]
LGTM, thank you!

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test Windows

**Comment by CodaFi** [comment]
@swift-ci smoke test Windows

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Pull request merged**